File: backend/internal/handler/dto/plaza.go
package dto

import "time"

// PlazaCurrencyMetaDTO carries the data needed to render CNY ⇄ USD toggle on
// the plaza pages. The `balance_recharge_multiplier` is sourced from the
// existing payment configuration: 1 CNY recharge → multiplier USD credited.
//
// model_native is always "USD" (LiteLLM-derived prices are USD); plan_native
// is always "CNY" (operator-priced subscription plans).
type PlazaCurrencyMetaDTO struct {
	BalanceRechargeMultiplier float64 `json:"balance_recharge_multiplier"`
	ModelNative               string  `json:"model_native"`
	PlanNative                string  `json:"plan_native"`
}

// PlazaImagePricesDTO encapsulates the three-tier pricing for image generation
// models (1K / 2K / 4K), in USD per image.
type PlazaImagePricesDTO struct {
	Tier1K float64 `json:"tier_1k"`
	Tier2K float64 `json:"tier_2k"`
	Tier4K float64 `json:"tier_4k"`
}

// PlazaModelRowDTO represents one (group, model) row on the public model plaza.
//
// The `type` field is either "token" or "image". For "token" rows the
// `*_per_mtok` fields are populated and the image-related fields are nil.
// For "image" rows the per-MTok fields are zero and the `*_image_prices`
// fields are populated.
type PlazaModelRowDTO struct {
	GroupID   int64  `json:"group_id"`
	GroupName string `json:"group_name"`
	Platform  string `json:"platform"`
	Model     string `json:"model"`
	Type      string `json:"type"`

	InputPricePerMTok      float64 `json:"input_price_per_mtok,omitempty"`
	OutputPricePerMTok     float64 `json:"output_price_per_mtok,omitempty"`
	SiteInputPricePerMTok  float64 `json:"site_input_price_per_mtok,omitempty"`
	SiteOutputPricePerMTok float64 `json:"site_output_price_per_mtok,omitempty"`

	// Token-only single-tier cache prices (USD per 1M tokens).
	//
	// Pointer semantics intentionally distinguish "absent" (nil → JSON omitted →
	// frontend renders "—") from "explicit zero" (&0 → JSON 0). Models that do
	// not declare `SupportsCacheBreakdown` and report zero cache pricing leave
	// these as nil so clients don't paint a misleading $0.
	CacheWritePricePerMTok     *float64 `json:"cache_write_price_per_mtok,omitempty"`
	CacheReadPricePerMTok      *float64 `json:"cache_read_price_per_mtok,omitempty"`
	SiteCacheWritePricePerMTok *float64 `json:"site_cache_write_price_per_mtok,omitempty"`
	SiteCacheReadPricePerMTok  *float64 `json:"site_cache_read_price_per_mtok,omitempty"`

	BaseImagePrices *PlazaImagePricesDTO `json:"base_image_prices,omitempty"`
	SiteImagePrices *PlazaImagePricesDTO `json:"site_image_prices,omitempty"`

	Multiplier      float64 `json:"multiplier"`
	DiscountPercent float64 `json:"discount_percent"`
}

// PlazaPlanGroupDTO is one group bundled by a subscription plan.
type PlazaPlanGroupDTO struct {
	GroupID        int64   `json:"group_id"`
	GroupName      string  `json:"group_name"`
	Platform       string  `json:"platform"`
	RateMultiplier float64 `json:"rate_multiplier"`
}

// PlazaPlanCardDTO represents one subscription-plan card on the public plaza.
type PlazaPlanCardDTO struct {
	ID            int64    `json:"id"`
	Name          string   `json:"name"`
	Description   string   `json:"description"`
	Price         float64  `json:"price"`
	OriginalPrice *float64 `json:"original_price,omitempty"`
	ValidityDays  int      `json:"validity_days"`
	ValidityUnit  string   `json:"validity_unit"`
	Features      string   `json:"features,omitempty"`
	// GroupID and the three flat fields below describe the primary group and are
	// kept for existing clients; Groups carries the full bundle.
	GroupID        int64               `json:"group_id"`
	GroupName      string              `json:"group_name"`
	Platform       string              `json:"platform"`
	RateMultiplier float64             `json:"rate_multiplier"`
	Groups         []PlazaPlanGroupDTO `json:"groups"`
	Models         []string            `json:"models"`
	ModelsOverflow int                 `json:"models_overflow"`
}

// PlazaModelsResponseDTO is the response payload for GET /api/v1/plaza/models.
type PlazaModelsResponseDTO struct {
	Rows         []PlazaModelRowDTO   `json:"rows"`
	CurrencyMeta PlazaCurrencyMetaDTO `json:"currency_meta"`
}

// PlazaPlansResponseDTO is the response payload for GET /api/v1/plaza/plans.
type PlazaPlansResponseDTO struct {
	Cards        []PlazaPlanCardDTO   `json:"cards"`
	CurrencyMeta PlazaCurrencyMetaDTO `json:"currency_meta"`
}

// PublicRechargePromoTierDTO 是公开充值赠送活动 DTO 的单个赠送档位。
type PublicRechargePromoTierDTO struct {
	MinAmount float64 `json:"min_amount"`
	BonusRate float64 `json:"bonus_rate"`
}

// PublicRechargePromoDTO 是 GET /api/v1/plaza/recharge-promo 的展示型 promo 体。
//
// 与 service.RechargePromo / checkout-info.recharge_promo 的差异（按 design D2）：
//   - 不含 `enabled`：能出现在响应里就意味着已启用且时间窗内
//   - 不含 `activity_id`：那是后端审计字段，对匿名前端无意义
//   - 多一个 `name`：作为运营文案位（首页 banner 标题），同时也回填到
//     checkout-info.recharge_promo.name 保持一致
type PublicRechargePromoDTO struct {
	Name       string                       `json:"name"`
	ValidFrom  *time.Time                   `json:"valid_from,omitempty"`
	ValidUntil *time.Time                   `json:"valid_until,omitempty"`
	Tiers      []PublicRechargePromoTierDTO `json:"tiers"`
	Version    string                       `json:"version"`
}

// PublicRechargePromoResponseDTO 是 GET /api/v1/plaza/recharge-promo 的整体响应。
//
// 当无生效活动时 `Promo` 为 nil（JSON 序列化为 `"promo": null`），
// 不使用 omitempty —— 前端始终能依靠 `promo` 字段是否为 null 来判断。
type PublicRechargePromoResponseDTO struct {
	Promo *PublicRechargePromoDTO `json:"promo"`
}

File: frontend/src/api/plaza.ts
/**
 * Public Pricing Plaza API endpoints.
 *
 * 这些端点公开未授权可访问，对应后端 GET /api/v1/plaza/{models,plans}。
 * 其中 `models` 接口的 query 参数已在后端校验：`q` ≤ 64 字符。
 */

import { apiClient } from './client'

// ==================== Types ====================

/** Currency context shared across both plaza endpoints. */
export interface PlazaCurrencyMeta {
  /** 1 CNY recharge → multiplier USD credited; used for CNY ⇄ USD on-the-fly conversion. */
  balance_recharge_multiplier: number
  /** Native currency for model rows (always "USD"). */
  model_native: 'USD'
  /** Native currency for plan rows (always "CNY"). */
  plan_native: 'CNY'
}

/** Three-tier image generation prices, all in USD per image. */
export interface PlazaImagePrices {
  tier_1k: number
  tier_2k: number
  tier_4k: number
}

/**
 * One row on the model plaza.
 *
 * `type === "token"` rows populate the `*_per_mtok` fields (USD per million tokens);
 * `type === "image"` rows populate the `*_image_prices` fields and leave token fields as 0.
 */
export interface PlazaModelRow {
  group_id: number
  group_name: string
  platform: string
  model: string
  type: 'token' | 'image'

  input_price_per_mtok?: number
  output_price_per_mtok?: number
  site_input_price_per_mtok?: number
  site_output_price_per_mtok?: number

  /**
   * Single-tier (5 minute) cache prices, USD per million tokens. Mirrors the
   * backend's `*float64`/`omitempty` semantics: the field is **omitted** when
   * the model has no cache pricing data and does not declare
   * `SupportsCacheBreakdown`. Frontend MUST render `—` for an absent value
   * and `$0` for an explicit zero (only emitted when the model declares
   * cache breakdown).
   */
  cache_write_price_per_mtok?: number
  cache_read_price_per_mtok?: number
  site_cache_write_price_per_mtok?: number
  site_cache_read_price_per_mtok?: number

  base_image_prices?: PlazaImagePrices
  site_image_prices?: PlazaImagePrices

  multiplier: number
  /** 整数百分比；例如 20 表示 20% off。 */
  discount_percent: number
}

/** One group bundled by a subscription plan. */
export interface PlazaPlanGroup {
  group_id: number
  group_name: string
  platform: string
  rate_multiplier: number
}

/** One subscription-plan card on the plaza. */
export interface PlazaPlanCard {
  id: number
  name: string
  description: string
  /** CNY (per `currency_meta.plan_native`). */
  price: number
  original_price?: number
  validity_days: number
  validity_unit: string
  features: string
  /** 主分组（groups 首元素），保留给徽标配色等需要单一代表分组的场景。 */
  group_id: number
  group_name: string
  platform: string
  /** Current default multiplier of the associated group. */
  rate_multiplier: number
  /** 套餐打包授予的全部分组；购买后每个分组各发放一条订阅。 */
  groups?: PlazaPlanGroup[]
  /** Up to 50 model names; if more exist, exposed via `models_overflow`. */
  models: string[]
  models_overflow: number
}

export interface PlazaModelsResponse {
  rows: PlazaModelRow[]
  currency_meta: PlazaCurrencyMeta
}

export interface PlazaPlansResponse {
  cards: PlazaPlanCard[]
  currency_meta: PlazaCurrencyMeta
}

// ==================== Public Recharge Promo ====================

/**
 * Single bonus tier on a public recharge campaign.
 *
 * Mirrors `dto.PublicRechargePromoTierDTO` 1-to-1.
 */
export interface PublicRechargePromoTier {
  /** Inclusive lower bound of the tier (CNY). */
  min_amount: number
  /** Multiplicative bonus rate, e.g. 0.05 means +5% extra USD credited. */
  bonus_rate: number
}

/**
 * Active recharge campaign payload returned by `GET /api/v1/plaza/recharge-promo`.
 *
 * Mirrors `dto.PublicRechargePromoDTO` exactly. The backend deliberately
 * **omits** `enabled` (presence implies enabled) and `activity_id`
 * (internal audit field) compared to the authenticated `checkout-info`
 * counterpart. The `version` string is the same dismiss-key contract
 * (`{id}:{updated_at_unix}`) used elsewhere.
 */
export interface PublicRechargePromo {
  /** Operator-authored display name; used as the banner title. Plain text — never inject as HTML. */
  name: string
  valid_from?: string
  valid_until?: string
  tiers: PublicRechargePromoTier[]
  version: string
}

/** Top-level response shape: `promo` is `null` when no campaign is active. */
export interface PublicRechargePromoResponse {
  promo: PublicRechargePromo | null
}

// ==================== Filter Params ====================

export interface PlazaModelsFilter {
  group_id?: number
  platform?: string
  q?: string
}

// ==================== API Surface ====================

export const plazaAPI = {
  /**
   * Fetch the model plaza rows.
   *
   * @param filter optional client-side filters; the backend interprets:
   *   - `group_id`: exact match
   *   - `platform`: case-insensitive exact match
   *   - `q`: case-insensitive substring on model name (≤ 64 chars; longer is rejected with 400)
   */
  async listModels(filter: PlazaModelsFilter = {}): Promise<PlazaModelsResponse> {
    const params: Record<string, unknown> = {}
    if (filter.group_id !== undefined) params.group_id = filter.group_id
    if (filter.platform) params.platform = filter.platform
    if (filter.q) params.q = filter.q
    const { data } = await apiClient.get<PlazaModelsResponse>('/plaza/models', { params })
    return data
  },

  /** Fetch the subscription-plan plaza cards. */
  async listPlans(): Promise<PlazaPlansResponse> {
    const { data } = await apiClient.get<PlazaPlansResponse>('/plaza/plans')
    return data
  },

  /**
   * Fetch the active recharge promo (anonymous-accessible).
   *
   * Resolves to `{ promo: null }` whenever no campaign is currently active —
   * including expired windows, empty tiers, and backend errors (the backend
   * silent-skips on its end so anonymous marketing surfaces never 5xx). The
   * response is intentionally HEADER-FREE: callers MUST NOT attach an
   * `Authorization` token, mirroring `/plaza/plans` behavior.
   */
  async getRechargePromo(): Promise<PublicRechargePromoResponse> {
    const { data } = await apiClient.get<PublicRechargePromoResponse>('/plaza/recharge-promo')
    return data
  },
}

export default plazaAPI
